File: AwaKinG/src/core/Camera.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using SharpDX;
using SharpDX.D3DCompiler;
using SharpDX.Direct3D;
using SharpDX.Direct3D11;
using SharpDX.DXGI;

using Buffer = SharpDX.Direct3D11.Buffer;
using Device = SharpDX.Direct3D11.Device;
using MapFlags = SharpDX.Direct3D11.MapFlags;
namespace AwaKinG.src.core
{
    public class Camera
    {
        public IInput IInput;

        public Vector3 Position {get { return _Pos; } private set { }}
        protected Vector3 _Pos;
        public Vector3 Look { get { return _Look; } private set { } }
        protected Vector3 _Look;
        protected Vector3 _Up;
        protected Vector3 _Right;

        public Matrix ViewMatrix;

        public Camera()
        {
            _Look = new Vector3(0.0f, 0.0f, 1.0f);
            _Right = new Vector3(1.0f, 0.0f, 0.0f);
            _Pos = new Vector3(0.0f, 0.0f, 0.0f);
        }
        public void Update()
        {
            
            float delta = 1.0f;
            if (IInput.LeftShift)
                delta = 4.0f;
            /*switch (_Camera_Type)
            {
                case Cam_type.Redactor:*/
        if (IInput.LeftControl)
        {
            Yaw(IInput.MX / 1000.0f);
            Pitch(IInput.MY / 1000.0f);
        }
        Walk(IInput.MZ * delta / 10.0f);
                 /*   break;
                case Cam_type.Free:
                    Yaw(IInput.MX / 1000.0f);
                    Pitch(IInput.MY / 1000.0f);
                    if (IInput.W)
                        Walk(1.1f * delta);
                    if (IInput.S)
                        Walk(-1.1f * delta);
                    if (IInput.A)
                        Strafe(-1.1f * delta);
                    if (IInput.D)
                        Strafe(1.1f * delta);
                    break;
            }
            if (IInput.R && IInput.LeftControl)
                _Camera_Type = Cam_type.Redactor;
            if (IInput.F && IInput.LeftControl)
                _Camera_Type = Cam_type.Free;
            */
            Frame();
        }
        private void Pitch(float angle)
        {
            Matrix T = Matrix.RotationAxis(_Right, angle);
            _Up = Vector3.TransformCoordinate(_Up, T);
            if (_Up.Y >= 0)
                _Look = Vector3.TransformCoordinate(_Look, T);
        }
        private void Yaw(float angle)
        {
            Matrix T = Matrix.RotationY(angle);
            _Right = Vector3.TransformCoordinate(_Right, T);
            _Look = Vector3.TransformCoordinate(_Look, T);
        }
        private void Walk(float units)
        {
            //if (type == CamType.First_Person)
            //_Pos += new Vector3(_Look.X, 0.0f, _Look.Z) * units;

            // if (type == CamType.Redaction)
            _Pos += _Look * units;
        }
        private void WalkLook(float units)
        {
            _Pos += _Look * units;
        }
        private void Strafe(float units)
        {
            /// case CamType.Redaction:
            _Pos += _Right * units;
            ///    break;
            /// case CamType.First_Person:
            ////  _Pos += new Vector3(_Right.X, 0.0f, _Right.Z) * units;
            ///   break;
        }
        private void Strafez(float units)
        {
            _Pos += _Up * units;
        }
        private void Frame()
        {
            Matrix V = new Matrix(0.0f);
            _Look = Vector3.Normalize(_Look);
            _Up = Vector3.Cross(_Look, _Right);
            _Up = Vector3.Normalize(_Up);

            //Right = Vector3.Cross(Up, Look);
            //Right = Vector3.Normalize(Right);

            float x = -Vector3.Dot(_Right, _Pos);
            float y = -Vector3.Dot(_Up, _Pos);
            float z = -Vector3.Dot(_Look, _Pos);

            V[0, 0] = _Right.X; V[0, 1] = _Up.X; V[0, 2] = _Look.X; V[0, 3] = 0.0f;
            V[1, 0] = _Right.Y; V[1, 1] = _Up.Y; V[1, 2] = _Look.Y; V[1, 3] = 0.0f;
            V[2, 0] = _Right.Z; V[2, 1] = _Up.Z; V[2, 2] = _Look.Z; V[2, 3] = 0.0f;
            V[3, 0] = x; V[3, 1] = y; V[3, 2] = z; V[3, 3] = 1.0f;
            ViewMatrix = V;
        }
    }
}

File: projects/AwaKinG_Engine/src/render/Camera.cpp
#include "Camera.h"

Camera::Camera() {
  _viewMatrix = XMMatrixIdentity();
  _position = XMFLOAT3(10.0f, 20.0f, 10.0f);
  _look = XMFLOAT3(0.0f, -1.0f, 0.0f);
  _up = XMFLOAT3(0.0f, 0.0f, 1.0f);
  _right = XMFLOAT3(1.0f, 0.0f, 0.0f);
}
void Camera::shutdown() {
}
XMMATRIX& Camera::getViewMatrix() {
  return _viewMatrix;
}
XMFLOAT3* Camera::getPosition() {
  return &_position;
}
XMFLOAT3* Camera::getLook() {
  return &_look;
}
void Camera::pitch(float angle) {
  XMMATRIX t_ = XMMatrixRotationAxis(XMLoadFloat3(&_right), angle);
  XMStoreFloat3(&_up, XMVector3TransformCoord(XMLoadFloat3(&_up), t_));

  if(_up.y >= 0)
    XMStoreFloat3(&_look, XMVector3TransformCoord(XMLoadFloat3(&_look), t_));
}
void Camera::yaw(float angle) {
  XMMATRIX t_ = XMMatrixRotationY(angle);
  XMStoreFloat3(&_right, XMVector3TransformCoord(XMLoadFloat3(&_right), t_));
  XMStoreFloat3(&_look, XMVector3TransformCoord(XMLoadFloat3(&_look), t_));
}
void Camera::walk(float units) {
  _position.x += _look.x*units;
  _position.y += _look.y*units;
  _position.z += _look.z*units;
}
void Camera::strafe(float units) {
  _position.x += _right.x*units;
  _position.y += _right.y*units;
  _position.z += _right.z*units;
}
void Camera::update() {
  XMVECTOR look_ = XMVector3Normalize(XMLoadFloat3(&_look));
  XMStoreFloat3(&_look, look_);
  XMVECTOR up_ = XMVector3Cross(look_, XMLoadFloat3(&_right));
  up_ = XMVector3Normalize(up_);
  XMStoreFloat3(&_up, up_);
  XMVECTOR right_ = XMVector3Cross(up_, look_);
  right_ = XMVector3Normalize(right_);
  XMStoreFloat3(&_right, right_);

  float x;
  XMStoreFloat(&x, XMVector3Dot(XMLoadFloat3(&_right), XMLoadFloat3(&_position))); x = -x;
  float y;
  XMStoreFloat(&y, XMVector3Dot(XMLoadFloat3(&_up), XMLoadFloat3(&_position))); y = -y;
  float z;
  XMStoreFloat(&z, XMVector3Dot(XMLoadFloat3(&_look), XMLoadFloat3(&_position))); z = -z;

  _viewMatrix._11 = _right.x; _viewMatrix._12 = _up.x;
  _viewMatrix._21 = _right.y; _viewMatrix._22 = _up.y;
  _viewMatrix._31 = _right.z; _viewMatrix._32 = _up.z;
  _viewMatrix._41 = x;				_viewMatrix._42 = y;

  _viewMatrix._13 = _look.x; _viewMatrix._14 = 0.0f;
  _viewMatrix._23 = _look.y; _viewMatrix._24 = 0.0f;
  _viewMatrix._33 = _look.z; _viewMatrix._34 = 0.0f;
  _viewMatrix._43 = z;			 _viewMatrix._44 = 1.0f;
}
void Camera::setPosition(XMFLOAT3& pos) {
  _position = pos;
}
void Camera::setLook(XMFLOAT3& look) {
  _look = look;
}